### PR TITLE
admission: temporary knob to disable AC for user requests 

### PR DESCRIPTION
**Pull request by andrewbaptist**
informs: #85641

Only leave admission control in place for elastic / bulk requests.

Release note: None

### CODE REVIEW

**Review by irfansharif — COMMENTED** [review]
I’ll happily concede to a second opinion here but IMO we should understand what the full extent of a principled fix needs to be before carving out backportable forms that also help the status quo. It may be that 22.2 and earlier benefit from just this change, but it’s something I’d rather us merge after we have the fuller fix. It makes for more churn in runbooks if anything.

**Comment by andrewbaptist** [comment]
I agree - I will not merge this yet. I started putting together a presentation that I wanted to go over tomorrow: https://docs.google.com/presentation/d/1N6N5fNAhS-Ec_fCv8rV2dJiVjvXlRb0-3r8PxTsnJPk/edit#slide=id.g170fb840c21_0_709 that covers this topic in a lot more detail.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90448)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/util/admission/work_queue.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFOzVg83Ta1IWi1tHcL:-NFOzVg9DPpCxZ3QxyZu:bahcp3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa7fb3abe86bfe5d9324521624dec1ead3f7b524/pkg/util/admission/work_queue.go#L57)):*
> ```Go
> // priority) work is subject to admission control. If it is set to true, then
> // user work will not be throttled by admission control but bulk work still will
> // be.
> ```

So when do we recommend this to be set to true?
Temporarily when they are running an import or index backfill?

Also please add a TODO to find a better backportable solution for v22.2.

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90448)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @sumeerbhola)

___
*[`pkg/util/admission/work_queue.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFOzVg83Ta1IWi1tHcL:-NFQFnvOD_IxUbv2Jdwg:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa7fb3abe86bfe5d9324521624dec1ead3f7b524/pkg/util/admission/work_queue.go#L57)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

So when do we recommend this to be set to true?
Temporarily when they are running an import or index backfill?

Also please add a TODO to find a better backportable solution for v22.2.
</blockquote></details>

Done

**Comment by sumeerbhola** [comment]
Let's wait on this in light of https://cockroachlabs.slack.com/archives/C01SRKWGHG8/p1666901963693889

**Comment by andrewbaptist** [comment]
I'm a little confused with the reluctance to merge this / backport to v22.2. Either we decide we didn't need it and we can ignore and rip it out later.  Alternatively places where we currently give guidance to disable admission control we change to use this setting instead. Long term as we discussed this should be deprecated, but it seems exceedingly unlikely that anything will go into v22.2.0 other than this patch. The KV support team would greatly appreciate having one more knob to turn to help manage AC induced degradation. We know that degradation can happen today based on the TPC-E testing.

File: pkg/util/admission/work_queue.go
Comment (irfansharif) [inline pkg/util/admission/work_queue.go:66]: Drop the .WithPublic(). We'll only use it selectively in escalations. It doesn't need to make its way into release notes or such, which brings with it more irrelevant backward compat considerations.

**Review by irfansharif — APPROVED** [review]
The reluctance is because it's a hack (though potentially useful like we've discussed). The reference to the other issue was because it's the only one I've seen discussed where we suspected this patch would've helped, but we know now it wouldn't. I'm ok merging this anyway.

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
+1 on reluctance reason.
Not sure why we are merging this given that we know that this is not a problem for v22.1 and given that we don't know how this will compare with turning on the follower pausing approach (or some backportable tweaks to it). But I'll leave that judgement call to KV folks. @tbg 

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-:-NFtOZN679cnrfUyTNGp:b3s3m5o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @irfansharif)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFtPMif3fc-kYKNQuGl):*
nit:
admission: add temporary knob to disable AC for user requests
___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFtPWy71Z6CiXC63Rly):*
Do we want a release note for non-public cluster settings?
___
*[`pkg/util/admission/work_queue.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFOzVg83Ta1IWi1tHcL:-NFtP-XQ0VIu1P8IlpZf:b-w6id78) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa7fb3abe86bfe5d9324521624dec1ead3f7b524/pkg/util/admission/work_queue.go#L57)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Don't see a mention of backportable solution for v22.2.
This should also have big warnings around it like:

It should ONLY be used reactively in cases ... 
___
*[`pkg/util/admission/work_queue.go` line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NEveshP8K8EVDzE766y-r2-66:-NFtPKPp3YCdhi0smLjR:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/06409362ecd891007df30f8c0727f4d23db860ac/pkg/util/admission/work_queue.go#L66)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Drop the .WithPublic(). We'll only use it selectively in escalations. It doesn't need to make its way into release notes or such, which brings with it more irrelevant backward compat considerations.
</blockquote></details>

+1

**Review by andrewbaptist — COMMENTED** [review]
We know how this compares with the follower pausing approach from Irfan's testing. We have both sets of graphs for the TPC-E index backfill. This patch performs significantly better and does not have any of the durability risks that the follower writes approach does. This is a much less risky change than that. I'm a little confused why we want to leave the "disable admission control" public and without depreciation plans but this much safer alternative has all these red flag warnings about it. I would much rather customers use this flag than disable all of AC. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-:-NFtXAyD3llok2c8Gi6S:b-822e64)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @irfansharif, and @sumeerbhola)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFtPMif3fc-kYKNQuGl):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit:
admission: add temporary knob to disable AC for user requests
</blockquote></details>

Done (added to the first line of description since it went over the char limit otherwise)
___
*[`-- commits` line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFtPWy71Z6CiXC63Rly):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Do we want a release note for non-public cluster settings?
</blockquote></details>

Done, removed release note
___
*[`pkg/util/admission/work_queue.go` line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NEveshP8K8EVDzE766y-r2-66:-NFtYXwv9l6UYbrh5KN3:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/06409362ecd891007df30f8c0727f4d23db860ac/pkg/util/admission/work_queue.go#L66)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done

**Review by irfansharif — APPROVED** [review]
> Not sure why we are merging this given that we know that this is not a problem for v22.1

I'm reading the above correctly this was more for 22.2.  

> But I'll leave that judgement call to KV folks.

I have a comparatively low bar for things getting merged (perhaps to a fault), especially if we have experimental intuition that it could help operationalize AC better. Going forward if I'm paged for index backfill related performance isolation issues, I'll recommend (a) the follower pausing thing first, (b) perhaps increasing the L0 file count/sublevel threshold next to see if I can stave off IO token exhaustion further, and (c) finally this one to just aggressively clamp down on AddSST/BulkNormalPri IO load. There might be a more direct knob in the index backfiller code that I'm unaware of. In any case, this feels ok to merge given all the caveats we're acknowledging above.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-:-NFuAO8K5LP8xQSPUmkZ:biaqz3u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @sumeerbhola)

**Comment by andrewbaptist** [comment]
bors r=irfansharif

**Comment by irfansharif** [comment]
This bors check is a nuisance, trying something.

bors r+

**Review by sumeerbhola — DISMISSED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90448)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @irfansharif)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFtPMif3fc-kYKNQuGl):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Done (added to the first line of description since it went over the char limit otherwise)
</blockquote></details>

nit: I am concerned that folks will read this title and draw the wrong conclusion.
Lines can overflow. Or shorten to:
admission: temporary knob to disable AC for user requests

if that does not fit, feel free to drop the temporary.
The title as is asserts that we have disabled AC for user requests, which is incorrect.
___
*[`pkg/util/admission/work_queue.go` line 57 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFOzVg83Ta1IWi1tHcL:-NFuPaUs0huWJzW3jpa1:bsoodxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa7fb3abe86bfe5d9324521624dec1ead3f7b524/pkg/util/admission/work_queue.go#L57)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Don't see a mention of backportable solution for v22.2.
This should also have big warnings around it like:

It should ONLY be used reactively in cases ... 
</blockquote></details>

Reminder

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90448)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @irfansharif)

___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFuQNaK5qegKFHEQfhR):*
nit:
s/Disable/disable/

**Comment by sumeerbhola** [comment]
btw, it didn't let me remove myself as a reviewer

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90448)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @sumeerbhola)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFtPMif3fc-kYKNQuGl):*
> temporary knob to disable AC for user requests
Done
___
*[`-- commits` line 2 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/90448#-NFuQNaK5qegKFHEQfhR):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit:
s/Disable/disable/
</blockquote></details>

Done

**Comment by andrewbaptist** [comment]
blathers backport 22.1

**Pull request merged**
